user/timepix3/module/src/Timepix3Event2StdEventConverter.cc

```diff
@@ -129,16 +130,30 @@ bool Timepix3RawEvent2StdEventConverter::Converting(eudaq::EventSPC ev, eudaq::S
         // The data is shifted 16 bits to the right, then 44 to the left in order to match the timestamp format (net 28 left)
         m_syncTime = (m_syncTime & 0x00000FFFFFFFFFFF) + ((pixdata & 0x00000000FFFF0000) << 28);
 
-        if(!m_clearedHeader && (m_syncTime / 4096 / 40) < 6000000) {
-          m_clearedHeader = true;
+        if(m_clearedHeader==0 && (m_syncTime / 4096 / 40) < 6000000) { // < 6sec
+          EUDAQ_INFO("Detected T0 signal in event: " + std::to_string(ev->GetEventNumber()));
+          m_clearedHeader++;
+
+        // From SPS data we know that even though pixel timestamps are not perfectly chronological, they are not more
+        // than "mixed up by -20us". At DESY, this is hardly (ever?) the case due to the lower occupancies.
+        // Hence, if the current timestamp is more than 20us earlier than the previous timestamp, we can assume that
+        // a 2nd T0 has occured. Take 100us with some safety margin.
+        // This implies we cannot detect a 2nd T0 within the first 100us after the initial T0. But did this ever happen?
+        } else if (m_syncTime < m_syncTime_prev - (50 * 4096 * 40)) {
+          m_clearedHeader++;
         }
+        m_syncTime_prev = m_syncTime;
       }
     }
 
     // Sometimes there is still data left in the buffers at the start of a run. For that reason we keep skipping data until
-    // this "header" data has been cleared, when the heart beat signal starts from a low number (~few seconds max)
-    if(!m_clearedHeader) {
+    // this "header" data has been cleared, when the heart beat signal starts from a low number (~few seconds max).
+    // To detect a possible second T0, we have no better gauge than the same criterion:
+    // Comparing the timestamp to the previous timestamp (see above).
+    if(m_clearedHeader == 0) {
         continue;
+    } else if(m_clearedHeader > 1) {
+        throw DataInvalid("Detected 2nd T0 signal in event: " + std::to_string(ev->GetEventNumber()));
```

Comment:
A bit dangerous this mesaage since it will be caught and printed by Corry - so it looks like sth happened in "Corry Event X" whereas it happened in "EUDAQ2 tIMEPIX3 eVENT x"

Reply:
Good point. In addition, I noticed that `ev->GetEventNumber()` always returns 0, so that's not a good message anyway.
However, I don't quite understand why it's always zero (even when it happens at 163sec or so)...

Reply:
The event number doesn't have to be set. One more of them many :nauseated_face: features of the EUDAQ event classes

Reply:
See https://github.com/simonspa/eudaq/blob/master/main/lib/core/include/eudaq/Event.hh#L113-L115

Reply:
Good to know. Then I'll simplify the message to "Detected second T0" to avoid confusion - I don't have access to the previous timestamp here any more and the time when it happens is printed in Corry anyway, so that would only double the output info.